runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,403 @@
+    }
+  }
+
+  /**
+   * The heart of splittable {@link DoFn} execution: processes a single (element, restriction) pair
+   * by creating a tracker for the restriction and checkpointing/resuming processing later if
+   * necessary.
+   *
+   * <p>TODO: This uses deprecated OldDoFn since DoFn does not provide access to state/timer
+   * internals. This should be rewritten to use the <a href="https://s.apache.org/beam-state">State
+   * and Timers API</a> once it is available.
+   */
+  private static class ProcessFn<
+          InputT, RestrictionT, OutputT, TrackerT extends RestrictionTracker<RestrictionT>>
+      extends OldDoFn<KeyedWorkItem<String, KV<InputT, RestrictionT>>, OutputT>
+      implements OldDoFn.RequiresWindowAccess {
+    /**
+     * The state cell containing a watermark hold for the output of this {@link DoFn}. The hold is
+     * acquired during the first {@link DoFn.ProcessElement} call for this element and restriction,
+     * and is released when the {@link DoFn.ProcessElement} call returns {@link
+     * DoFn.ProcessContinuation#stop}.
+     *
+     * <p>A hold is needed to avoid letting the output watermark immediately progress together with
+     * the input watermark when the first {@link DoFn.ProcessElement} call completes.
+     *
+     * <p>The hold is updated with the future output watermark reported by ProcessContinuation.
+     */
+    private static final StateTag<Object, WatermarkHoldState<GlobalWindow>> watermarkHoldTag =
+        StateTags.makeSystemTagInternal(
+            StateTags.<GlobalWindow>watermarkStateInternal(
+                "hold", OutputTimeFns.outputAtLatestInputTimestamp()));
+
+    /**
+     * The state cell containing a copy of the element. Written during the first {@link
+     * DoFn.ProcessElement} call and read during subsequent calls in response to timer firings, when
+     * the original element is no longer available.
+     */
+    private final StateTag<Object, ValueState<WindowedValue<InputT>>> elementTag;
+
+    /**
+     * The state cell containing a restriction representing the unprocessed part of work for this
+     * element.
+     */
+    private StateTag<Object, ValueState<RestrictionT>> restrictionTag;
+
+    private final DoFn<InputT, OutputT> fn;
+    private transient DoFnInvoker<InputT, OutputT> invoker;
+
+    ProcessFn(
+        DoFn<InputT, OutputT> fn,
+        Coder<InputT> elementCoder,
+        Coder<? extends BoundedWindow> windowCoder) {
+      this.fn = fn;
+      elementTag =
+          StateTags.value("element", WindowedValue.getFullCoder(elementCoder, windowCoder));
+    }
+
+    @Override
+    public void setup() throws Exception {
+      invoker = DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn);
+      restrictionTag =
+          StateTags.value("restriction", invoker.<RestrictionT>invokeGetRestrictionCoder());
+    }
+
+    @Override
+    public void processElement(final ProcessContext c) {
```

Comment:
This `processElement` method is long. It would be good to split it up into some helper functions or even a helper class to make the different steps clearer.